Refactor declaring repositories into core page and cheat sheet (#22943)
gradle/gradle: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc

[[declaring-repositories]]
= Declare Repositories

Gradle allows you to define **repositories**: sources for dependency downloads.
During a build, Gradle searches repositories to resolve dependencies.
You can specify repositories as locations in a file system, or as URLs.
Gradle can resolve dependencies from one or many repositories.

[TIP]
====

> Can't find the information you're looking for?

Author: Interesting. I can add a bullet point list to each of these. My only concern is length -- bulleted lists in particular take up a LOT of space in admonitions, and it really breaks up the page. It can also push content below the fold, especially on smaller screens. Given that concern, would you still like to add the bulleted list?

> Can't find the information you're looking for?

Reviewer: Your right that would break it up too much.  Perhaps adding just a brief description like "for advanced content about repositories such as authentication, remote protocols and local configurations"

[[sec:declaring_multiple_repositories]]
== Declare Multiple Repositories

You an declare more than one repository within a single build.
This can come in handy when certain dependencies are only available in one repository.

Gradle uses the following rules to decide which repository to use to download each dependency:

- Gradle searches repositories in the order that they are declared.
- When Gradle finds a dependency, it attempts to download all artifacts for that dependency from that same repository.

The following build script example declares multiple repositories:

====
include::sample[dir="snippets/dependencyManagement/declaringRepositories-multipleRepositories/groovy",files="build.gradle[tags=multiple-repositories]"]
include::sample[dir="snippets/dependencyManagement/declaringRepositories-multipleRepositories/kotlin",files="build.gradle.kts[tags=multiple-repositories]"]
====

[NOTE]
====
For more information about dependency resolution with multiple repositories, see <<dependency_resolution.adoc#sec:how-gradle-downloads-deps, dependency downloads>>.
====

[[sec:declaring_public_repository]]

> == Declare a Repository with Shorthand Notation

Author: We don't have an established pattern beyond "title case". Many of our existing titles are either not titlecase or MLA, so I opted for MLA for something-closer-to-consistency.

=== Declare the Maven Central Repository

The Maven Central repository hosts open source libraries for Java projects.
To add the link:https://repo.maven.apache.org/maven2/[Maven Central repository] to your build, add the following to your build script:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-central]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-central]"]
====

[[sub:maven_google]]
=== Declare the Google Maven Repository

The Google Maven repository hosts Android-specific artifacts, including the Android SDK.
To add the link:https://maven.google.com/[Google Maven repository] to your build, add the following to your build script:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-google]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-google]"]
====

[[sub:flat_dir_resolver]]
== Declare a Flat Directory Repository

Some projects store dependencies on a shared drive or as part of the project source code instead of a binary repository product.

> Gradle calls this a directory repository.

Author: I will leave this one for the technical review. The original document sometimes omits "flat" or uses it in a parenthetical. I eliminated it here because "directory repository" feels like enough of a description and generally fewer words == better, more readable doc. However, the API does use the term `flatDir`, so maybe we should emphasize "flat" more.

[[sec:maven_repo]]
== Declare a Maven Repository

You can declare Maven repositories by URL.
To add a Maven repository to your build, add the following to your build script:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-like-repo]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-like-repo]"]
====

Maven POM metadata can reference external repositories.
Gradle ignores these additional repositories to ensure the reproducibility and security of Gradle builds.

[[sub:custom-maven-repo]]
=== Declare a Composite Maven Repository

Sometimes repositories publish metadata and artifacts in separate locations.
Gradle calls these *composite repositories*.
To add a composite Maven repository to your build, add the following to your build script:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-like-repo-with-jar-repo]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-like-repo-with-jar-repo]"]
====

Gradle automatically queries the base `url` location for metadata and artifacts.

> If artifacts can't be found at the base URL, Gradle checks for artifacts at the extra `artifactUrls`.

Author: Longer sentences generally impede readability. I feel that these thoughts are complex enough that they each deserve sentences of their own.

[[sub:accessing_secured_maven_repositories]]
=== Access an Authenticated Maven Repository

You can specify credentials for Maven repositories secured by authentication.

> Specify credentials for Maven repositories secured by basic authentication with link:{javadocPath}/org/gradle/api/credentials/PasswordCredentials.html[PasswordCredentials]:

Author: I don't think this is correct?

[[sub:declare-a-release-or-snapshot-maven-repository]]
=== Declare a Release or Snapshot Maven Repository

Maven repositories sometimes contain either release artifacts or snapshot artifacts, but not both.
You can filter Maven repositories based on the artifact type with the `mavenContent` block:

====
include::sample[dir="snippets/dependencyManagement/declaringRepositories-filtering/groovy",files="build.gradle[tags=repository-snapshots]"]
include::sample[dir="snippets/dependencyManagement/declaringRepositories-filtering/kotlin",files="build.gradle.kts[tags=repository-snapshots]"]
====

[[sec:ivy_repositories]]
== Declare an Ivy Repository

You can declare Ivy repositories by URL.
To add an Ivy repository to your build, add the following to your build script:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo]"]
====

[[sub:defining_a_named_layout_for_an_ivy_repository]]
=== Specify a Named Layout for an Ivy Repository

> You can specify that your repository conforms to the Ivy or Maven default layout with a named layout.
> To add an Ivy repository with a named layout to your build, add the following to your build script:

Author: ```suggestion
You can specify that your repository conforms to the Ivy or Maven default layout with a named layout.
To add an Ivy repository with a named layout to a build, add the following to your build script:
```
Good point. Removed one.

include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-maven-layout]"]
====

Gradle accepts the following layout values for Ivy repositories:

* `gradle`
* `maven`
* `ivy`

By default, Gradle uses the `gradle` layout.
For more information about these layouts, see link:{groovyDslPath}/org.gradle.api.artifacts.repositories.IvyArtifactRepository.html#org.gradle.api.artifacts.repositories.IvyArtifactRepository:layout(java.lang.String)[IvyArtifactRepository.layout(java.lang.String)].

[[sub:defining_custom_pattern_layout_for_an_ivy_repository]]
=== Specify a Custom Pattern Layout for an Ivy Repository

To declare an Ivy repository with a non-standard layout, add a *pattern layout* to the repository declaration:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo-with-pattern-layout]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-pattern-layout]"]
====

Some repositories store Ivy metadata and artifacts in different locations. 
You can define separate patterns to fetch Ivy metadata and artifacts from these locations.

> Use `artifact` to specify artifact location patterns, and `ivy` to specify Ivy metadata location patterns.

Author: I prefer to keep this comma here since this is a list of separate-but-related thoughts.

You can add multiple locations for both artifacts and metadata.
Gradle uses these patterns in the order that they are defined.
The following example shows a build script that declares multiple patterns:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo-with-custom-pattern]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-custom-pattern]"]
====

==== Specify a Maven Style Custom Pattern Layout

> Some Ivy repositories use Maven style for organisation layouts, where forward slashes replace dots as separators.

Author: This concept uses the British form in the API, so I've used the British form here for consistency.

Normally I would use American.

To use Maven style organization layouts in a repository, assign `m2compatible` a value of "true" in the `patternLayout` block:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo-with-m2compatible-layout]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-m2compatible-layout]"]
====

[[sub:accessing_secured_ivy_repositories]]
=== Access an Authenticated Ivy Repository

You can specify credentials for Ivy repositories secured by basic authentication:

====
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=authenticated-ivy-repo]"]
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=authenticated-ivy-repo]"]
====

For additional authentication options, see <<repositories_cheat_sheet.adoc#sec:supported_transport_protocols,supported transport protocols>>.

[[sec:repository-content-filtering]]
== Filter Repository Contents

Gradle exposes an API to control which dependencies your project can fetch from a repository.
Common use cases include:

> * performance (e.g. you know a dependency will never be found in a specific repository)

Reviewer: question: I thought usage was `e.g.,` but divert to your knowledge there

Author: It's complicated, but I find it's helpful to think of it as "example given".

* uses an inclusion clause to fetch only dependencies produced by a specific organization from a maven repository defined with a custom URL
* uses an exclusion clause to exclude dependencies produced by that same organization from the Maven Central repository defined with built-in shorthand notation

====
include::sample[dir="snippets/dependencyManagement/declaringRepositories-filtering/groovy",files="build.gradle[tags=repository-filter]"]
include::sample[dir="snippets//dependencyManagement/declaringRepositories-filtering/kotlin",files="build.gradle.kts[tags=repository-filter]"]
====

By default, repositories include everything and exclude nothing.
Adding an include or exclude clause has the following side effects:

* If you declare only includes, Gradle includes *only dependencies that match the includes*.
* If you declare only excludes, Gradle includes *everything except dependencies that match the excludes*.
* If you declare both includes and excludes, Gradle includes *only dependencies that match the includes that also do not match the excludes*.

You can filter by _group_, _module_ or _version_, either with strict string matching or regular expressions.

> You can filter versions by version range, using <<single_versions.adoc#single-version-declarations,the Gradle version format>>.

Reviewer: nitpick:
```suggestion
You can filter versions by version range using <<single_versions.adoc#single-version-declarations,the Gradle version format>>.
```

> You can filter by _group_, _module_ or _version_, either with strict string matching or regular expressions.
> You can filter versions by version range using <<single_versions.adoc#single-version-declarations,the Gradle version format>>.
> You can also filter by configuration name or attributes.

Author: ```suggestion
You can filter by _group_, _module_ or _version_, configuration name, or attributes, either with strict string matching or regular expressions.
Version filtering supports version ranges using <<single_versions.adoc#single-version-declarations,the Gradle version format>>.
```
Fix for https://github.com/gradle/gradle/pull/22943/files#r1038346574

> You can also filter by configuration name or attributes.

Reviewer: thought: I do not like the "you can" pattern here. Less words are better and wonder if this would be best in a list format?

Author: Fixed: https://github.com/gradle/gradle/pull/22943/files#r1038471110
